File: pocketr-java.api/src/main/java/com/decrux/pocketr/api/services/reporting/GenerateReport.java
package com.decrux.pocketr.api.services.reporting;

import com.decrux.pocketr.api.entities.db.auth.User;
import com.decrux.pocketr.api.entities.dtos.AccountBalanceSummaryDto;
import com.decrux.pocketr.api.entities.dtos.AccountBalanceTimeseriesDto;
import com.decrux.pocketr.api.entities.dtos.MonthlyExpenseDto;
import java.time.LocalDate;
import java.time.YearMonth;
import java.util.List;
import java.util.UUID;

public interface GenerateReport {

    List<MonthlyExpenseDto> getMonthlyExpenses(User user, YearMonth period, String mode, UUID householdId);

    List<AccountBalanceSummaryDto> getAllAccountBalances(User user, LocalDate asOf);

    AccountBalanceTimeseriesDto getBalanceTimeseries(UUID accountId, LocalDate dateFrom, LocalDate dateTo, User user);
}

File: pocketr-api/src/main/kotlin/com/decrux/pocketr_api/services/reporting/GenerateReport.kt
package com.decrux.pocketr_api.services.reporting

import com.decrux.pocketr_api.entities.db.auth.User
import com.decrux.pocketr_api.entities.dtos.AccountBalanceSummaryDto
import com.decrux.pocketr_api.entities.dtos.AccountBalanceTimeseriesDto
import com.decrux.pocketr_api.entities.dtos.MonthlyExpenseDto
import java.time.LocalDate
import java.time.YearMonth
import java.util.UUID

interface GenerateReport {
    fun getMonthlyExpenses(
        user: User,
        period: YearMonth,
        mode: String,
        householdId: UUID?,
    ): List<MonthlyExpenseDto>

    fun getAllAccountBalances(
        user: User,
        asOf: LocalDate,
    ): List<AccountBalanceSummaryDto>

    fun getBalanceTimeseries(
        accountId: UUID,
        dateFrom: LocalDate,
        dateTo: LocalDate,
        user: User,
    ): AccountBalanceTimeseriesDto
}
